Document/0x06f-Testing-Local-Authentication.md

```diff
@@ -9,26 +9,196 @@ As a result only requirement "4.6	Biometric authentication, if any, is not event
+query[kSecAttrAccount as String] = "My Name" as CFString
+query[kSecAttrLabel as String] = "com.me.myapp.password" as CFString
+query[kSecUseOperationPrompt as String] = "Please, pass authorisation to enter this area" as CFString
+
+// 2. get item
+var queryResult: AnyObject?
+let status = withUnsafeMutablePointer(to: &queryResult) {
+    SecItemCopyMatching(query as CFDictionary, UnsafeMutablePointer($0))
+}
+
+if status == noErr {
+    let password = String(data: queryResult as! Data, encoding: .utf8)!
+    // successfully received password
+} else {
+    // authorization not passed
+}
+```
+###### Objective-C
+```
+// 1. define query
+NSDictionary *query = @{(__bridge id)kSecClass: (__bridge id)kSecClassGenericPassword,
+    (__bridge id)kSecReturnData: @YES,
+    (__bridge id)kSecAttrAccount: @"My Name1",
+    (__bridge id)kSecAttrLabel: @"com.me.myapp.password",
+    (__bridge id)kSecUseOperationPrompt: @"Please, pass authorisation to enter this area" };
+
+// 2. get item
+CFTypeRef queryResult = NULL;
+OSStatus status = SecItemCopyMatching((__bridge CFDictionaryRef)query, &queryResult);
+
+if (status == noErr){
+    NSData *resultData = ( __bridge_transfer NSData *)queryResult;
+    NSString *password = [[NSString alloc] initWithData:resultData encoding:NSUTF8StringEncoding];
+    NSLog(@"%@", password);
+} else {
+    NSLog(@"Something went wrong");
+}
+```
+#### Without Source Code
+Usage of frameworks in app can be detected by analyzing app binary's list of shared dynamic libraries by running:
+ `$ otool -L <AppName>.app/<AppName>`
+
+If `LocalAuhentication.framework` was used in app, output will contain both of following lines (remember that `LocalAuhentication.framework` uses `Security.framework` under the hood), if `Security.framework` - only second one:
+```
+/System/Library/Frameworks/LocalAuthentication.framework/LocalAuthentication
+/System/Library/Frameworks/Security.framework/Security
+```
 
 #### Dynamic Analysis
 
--- TODO [Describe how to test for this issue "Testing Biometric Authentication" by running and interacting with the app. This can include everything from simply monitoring network traffic or aspects of the app’s behavior to code injection, debugging, instrumentation, etc.] --
-
-#### Remediation
+On a running app usage of TouchID authentication is quite obvious: at appropriate moment system-style alert asks user to put her finger on sensor or fallback to password (if allowed):
+![TouchID authentication UI](/Images/Chapters/0x6f/biometric_auth_try_again.jpg)
+![TouchID authentication UI with fallback to passcode](/Images/Chapters/0x6f/biometric_auth_try_again.jpg)
 
-Example: https://www.raywenderlich.com/92667/securing-ios-data-keychain-touch-id-1password
--- TODO [Describe the best practices that developers should follow to prevent this issue "Testing Biometric Authentication".] --
+On a jailbroken device contents of Keychain can be dumped and items' parameters inspected.
```

Comment:
On a jailbroken device TouchID can be bypassed if only LocalAuthentication framework is implemented, by using Swizzler (https://github.com/vtky/Swizzler):

1. Settings->Swizzler
2. Enable "Inject Swizzler into Apps"
3. Enable "Log Everything to Syslog"
4. Enable "Log Everything to File"
5. Enter the submenu "iOS Frameworks"
6. Enable "LocalAuthentication"
7. Enter the submenu "Select Target Apps"
8. Enable the target app
9. Close the app and start it again
10. When the TouchID prompt shows click "cancel"
11. If the application flow continues without requiring the touchID then the bypass has worked